compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -4813,6 +4813,11 @@ object Types {
   /** The type of an import clause tree */
   case class ImportType(expr: Tree) extends UncachedGroundType
 
+  /** Sentinal for typed export clauses */
+  @sharable case object ExportType extends CachedGroundType {
```

Comment:
It seems that we could also use ImportType instead of adding a new type node, it wouldn't be very useful but ImportType itself isn't very useful either, it would just be nice for symmetry. Or is there some reason we couldn't use ImportType?

Reply:
I tried using an import type but supplying the `expr` of the Export tree, but then they would fail in Ycheck because I guess the types had changed identity despite being structurally the same

Reply:
otherwise we can generate a symbol for each export node and use a term ref to that

Reply:
what was the failure in Ycheck? If we process Import and Export in the same way it seems weird that one would work but not the other

Reply:
> what was the failure in Ycheck? If we process Import and Export in the same way it seems weird that one would work but not the other

I'd still be interested in understanding what's going on here, otherwise LGTM.

Reply:
if i use an `ImportType` instead:
```scala
  def assignType(tree: untpd.Export)(using Context): Export =
    tree.withType(ImportType(tree.expr.asInstanceOf[tpd.Tree].ensuring(_.tpe != null)))
```
then we crash in ycheck due to a subtype check fail:
```
Fatal compiler crash when compiling: tests/run-macros/exports:
assertion failed: Types differ
Original type : ImportType(Select(This(Ident($anon)),delegate))
After checking: ImportType(Select(This(Ident($anon)),delegate))
Original tree : export this.delegate._
After checking: export this.delegate._
Why different :
             Subtype trace:
  ==> ImportType(Select(This(Ident($anon)),delegate)) <:< ImportType(Select(This(Ident($anon)),delegate))  
    ==> ImportType(Select(This(Ident($anon)),delegate)) <:< ImportType(Select(This(Ident($anon)),delegate)) recur 
    <== ImportType(Select(This(Ident($anon)),delegate)) <:< ImportType(Select(This(Ident($anon)),delegate)) recur  = false
  <== ImportType(Select(This(Ident($anon)),delegate)) <:< ImportType(Select(This(Ident($anon)),delegate))   = false
scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
dotty.tools.dotc.transform.TreeChecker$Checker.typedUnadapted(TreeChecker.scala:346)
dotty.tools.dotc.typer.Typer.typed(Typer.scala:2615)
dotty.tools.dotc.typer.Typer.typed(Typer.scala:2619)
dotty.tools.dotc.transform.TreeChecker$Checker.typed(TreeChecker.scala:317)

```

Reply:
The difference is that an `Import` tree is always assigned a singleton type ref to its import symbol, which has an info of `ImportType`

Reply:
Ah, so it's different from Import because Imports have a symbol and use we can use that to get a termRef? So either we also create a symbol for Export nodes or we keep the empty ExportType as is. For symmetry I would be tempted to also create symbols for export nodes, on the other hand I don't know if that would be useful (are import symbols useful at all?)

Reply:
So import symbols are used in typer to lookup members that come from an import:
https://github.com/lampepfl/dotty/blob/21aee87606e08b2bdbbda72863004f4449937c65/compiler/src/dotty/tools/dotc/typer/Typer.scala#L196

Reply:
I thought the best way to capture the exported symbols would be to record them at the time that the namer generates the export forwarders. If you inline a template with an export clause then the symbols no longer correspond to the originally recorded forwarders, as new symbols are generated.

Reply:
my concern with not recomputing forwarders at a later phase is if it could be possible for extra members to be added to the prefix that do not then have corresponding forwarders

Reply:
Then if you are recomputing forwarders at the point of unpickling, i then don't know if we want to be forcing the members of the enclosing scope to check that the signatures match up

Reply:
> Then if you are recomputing forwarders at the point of unpickling

We should never do that, the forwarders are pickled so this shouldn't be necessary.

Reply:
Current implementation in the commit pickles the signed names of the definitions instead, which can be used to recover the symbols as shown in the test

Reply:
Unfortunately, this also does not work because the signature will change for an inlined method referring to an inlined class

Reply:
I'll try a version that does not care about signatures

Reply:
its more safe now that to compare against the real forwarders you have the protection of checking for the Exported flag

Reply:
In the dotty meeting we decided to go with the sentinel type for now